CRM/Contribute/Form/ContributionRecur.php

```diff
@@ -0,0 +1,535 @@
+        "reset=1&cid={$this->_contactID}&selectedChild=contribute-recur"
+      ));
+      return NULL;
+    }
+
+    // get the submitted form values.
+    $submittedValues = $this->controller->exportValues($this->_name);
+
+    // get the required field value only.
+    $formValues = $submittedValues;
+    $ids = array();
+
+    if ($this->_contactID) {
+      $params['contact_id'] = $this->_contactID;
+    }
+
+    $params['currency'] = CRM_Contribute_Form_AbstractEditPayment::getCurrency($submittedValues);
+
+    $dates = array(
+      'start_date',
+      'end_date',
+      'cancel_date',
+      'next_sched_contribution_date',
+    );
+
+    foreach ($dates as $d) {
+      $params[$d] = CRM_Utils_Date::processDate($formValues[$d], $formValues[$d . '_time'], TRUE);
+    }
+
+    if (empty($this->_id)) {
+      $params['create_date'] = CRM_Utils_Date::processDate(date('Y-m-d'));
+    }
+
+    if ($this->_action & CRM_Core_Action::ADD) {
+      $params['next_sched_contribution_date'] = $params['start_date'];
+    }
+    $fields = array(
+      'payment_processor_id',
+      'processor_id',
+      'trxn_id',
+      'financial_type_id',
+      'amount',
+      'payment_instrument_id',
+      'frequency_interval',
+      'frequency_unit',
+      'cycle_day',
+      'contribution_status_id',
+    );
+    foreach ($fields as $f) {
+      $params[$f] = CRM_Utils_Array::value($f, $formValues);
+    }
+
+    $params['id'] = $this->_id;
+
+    // build custom data getFields array
+    $customFields = CRM_Core_BAO_CustomField::getFields('ContributionRecur', FALSE, FALSE, NULL, NULL, TRUE);
+    $params['custom'] = CRM_Core_BAO_CustomField::postProcess($_POST,
+      $this->_id,
+      'ContributionRecur'
+    );
+
+    $contributionRecur = CRM_Contribute_BAO_ContributionRecur::create($params);
+
+    // Link the recurring contribution with membership record, if selected
+    if ($this->_action == 1 && !empty($formValues['membership_id'])) {
+      civicrm_api3('Membership', 'create', array(
+        'id' => $formValues['membership_id'],
+        'contribution_recur_id' => $contributionRecur->id,
+      ));
+    }
+
+    // Move the recurring record
+    if (isset($submittedValues['move_recurring_record']) && $submittedValues['move_recurring_record'] == 1) {
+      self::moveRecurringRecord($submittedValues);
+    }
+
+    $session->replaceUserContext(CRM_Utils_System::url('civicrm/contact/view',
+      "reset=1&cid={$this->_contactID}&selectedChild=contribute-recur"
+    ));
+  }
+
+  public function displayStatusMessage($messageTitle, $message) {
+    $out = array(
+      'status' => 'fatal',
+      'content' => '<div class="messages status no-popup"><div class="icon inform-icon"></div>' . ts($message) . '</div>',
+    );
+    CRM_Core_Session::setStatus($message, ts($messageTitle), 'error');
+    CRM_Core_Page_AJAX::returnJsonResponse($out);
+  }
+
+  public function moveRecurringRecord($submittedValues) {
```

Comment:
Can this function (with a comment block) be moved to the BAO - we don't really want to put new functionity into forms. The membership_record would be passed in rather than having the BAO look in POST - but note that this is not secure code $membership_record = $_POST['membership_record']; - you should check with @totten for how to a recommendation (I don't personally know why you would need to get it from there - which makes it hard for me to comment on that - quickform would not return it to you if it is not a quickform element - but you probably DO need it in quickform if you want it submitted)